src/dmd/root/longdouble.d

```diff
@@ -6,29 +6,92 @@
  * https://github.com/dlang/dmd/blob/master/src/root/longdouble.d
  */
 
-// 80 bit floating point value implementation for LDC compiler targetting MSVC
+// 80-bit floating point value implementation if the C/D compiler does not support them natively
 
 module dmd.root.longdouble;
 
-enum hasNativeFloat80 = real.sizeof > 8;
-
-static if (hasNativeFloat80)
-{
+static if (real.sizeof > 8)
     alias longdouble = real;
-}
+else
+    alias longdouble = longdouble_soft;
 
-static if (!hasNativeFloat80):
+// longdouble_soft needed when building the backend with
+// Visual C or the frontend with LDC on Windows
+version(CRuntime_Microsoft):
 extern(C++):
 nothrow:
+@nogc:
 
-align(2) struct longdouble
+version(D_InlineAsm_X86_64)
+    version = AsmX86;
+else version(D_InlineAsm_X86)
+    version = AsmX86;
+else
+    static assert(false, "longdouble_soft not supported on this platform");
+
+bool initFPU()
 {
-nothrow:
+    version(D_InlineAsm_X86_64)
+    {
+        // set precision to 64-bit mantissa and rounding control to nearest
+        asm nothrow @nogc pure
```

Comment:
`nothrow` and `@nogc` are already declared above.  Is this redundant?  If so, I suggest choosing a coding style:  either add the attributes explicitly at each declaration, or group them under one attribution, but don't mix the two styles.

Reply:
Not confident enough for `@trusted`?

Reply:
> nothrow and @nogc are already declared above. Is this redundant?

These attributes have to be repeated for each asm statement. otherwise the compiler errors out.

Reply:
Seems like a bug.

Reply:
Before dmd 2.067, you couldn't write asm in pure/safe code: https://dlang.org/changelog/2.067.0.html#asm-attributes